src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,143 @@
+            tail = newNode;
+        }
+        size++;
+        return true;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index == size) {
+            add(value);
+            return;
+        }
+        checkIndex(index);
+        if (index == 0) {
+            Node<T> newNode = new Node<>(null, value, head);
+            head.prev = newNode;
+            head = newNode;
+        } else {
+            Node<T> currentNode = node(index);
+            Node<T> newNode = new Node<>(currentNode.prev, value, currentNode);
+            currentNode.prev.next = newNode;
+            currentNode.prev = newNode;
+        }
+        size++;
     }
 
     @Override
     public boolean addAll(List<T> list) {
-        return false;
+        for (T values : list) {
+            add(values);
+        }
+        return true;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return node(index).value;
+
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        T returnedValue = node(index).value;
+        node(index).value = value;
+        return returnedValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> removedValue = node(index);
+        return unlink(removedValue);
     }
 
     @Override
     public boolean remove(T object) {
+        for (int i = 0; i < size; i++) {
+            if (node(i).value == object || object != null && object.equals(node(i).value)) {
+                remove(i);
+                return true;
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
+        if (size == 0) {
+            return true;
+        }
         return false;
     }
+
+    private boolean checkIndex(int index) {
+        if (index >= size || index < 0) {
+            throw new IndexOutOfBoundsException("Index out of bound exception");
+        }
+        return true;
+    }
+
+    Node<T> node(int index) {
+        if (index < (size * 3 / 2)) {
+            Node<T> newNode = head;
+            for (int i = 0; i < index; i++) {
+                newNode = newNode.next;
+            }
+            return newNode;
+        } else {
+            Node<T> newNode = tail;
+            for (int i = size - 1; i > index; i--) {
+                newNode = newNode.prev;
+            }
+            return newNode;
+        }
+    }
+
+    private T unlink(Node<T> removedValue) {
+        if (removedValue.next == null) {
+            tail = removedValue.prev;
+        } else if (removedValue.prev == null) {
+            head = removedValue.next;
+        } else {
+            removedValue.prev.next = removedValue.next;
+            removedValue.next.prev = removedValue.prev;
+        }
```

Comment:
this is not a 100% correct implementation at the moment, you should unlink from both sides in each case

let’s try to refactor this a bit
create 2 variables for `nextNode` and `previousNode`

first check if `nextNode` is null - write `if-else` construction

then check if `previousNode` is null - write another `if-else` construction

let’s also set node’s value to null in the end